### PR TITLE
Allow results to work when pipeline spec is embedded ⛏️

### LINKED ISSUE #2463: Results require runAfter when embedding Pipeline spec
# Expected Behavior

Embedding a Pipeline spec into a PipelineRun should not change the behaviour of results.

# Actual Behavior

If you remove the `runAfter` clause from [this example](https://github.com/tektoncd/pipeline/blob/master/examples/v1beta1/pipelineruns/pipelinerun-results-with-params.yml), you get an error caused by the controller mistakenly starting both Tasks simultaneously.

```yaml
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  generateName: test-case
  name: test-case-run
spec:
  params:
  - name: prefix
    value: prefix
  pipelineSpec:
    params:
    - name: prefix
    tasks:
    - name: generate-suffix
      taskSpec:
        results:
        - name: suffix
        steps:
        - name: generate-suffix
          image: alpine
          script: |
            echo -n "suffix" > $(results.suffix.path)
    - name: do-something
      taskSpec:
        params:
        - name: arg
        steps:
        - name: do-something
          image: alpine
          script: |
            echo "$(params.arg)"
      params:
      - name: arg
       # this should be enough to imply to the controller that this Task should run
       # after generate-suffix, but with runAfter removed, the controller starts it at
       # the same time as generate-suffix
        value: "$(params.prefix):$(tasks.generate-suffix.results.suffix)"
```

# Steps to Reproduce the Problem

1. Apply the above yaml and look at the pipelinerun

# Additional Info

Interestingly this doesnt happen when not embedding the pipelinespec, e.g. this other example does not have the same problem https://github.com/tektoncd/pipeline/blob/master/examples/v1beta1/pipelineruns/pipelinerun-results.yaml

- Tekton Pipeline version: HEAD @ 1346656111decda59d4c36280b2cc828127a37e1

### PR DESCRIPTION
**Pull request by bobcatfish**
# Changes

Yesterday I thought I would just quickly update our existing build +
deploy pipelinerun example to use a git task and a kaniko task instead
of PipelineResources and I immediately started stumbling over some very
strange bugs.

I'm not sure I've gotten 100% to the bottm of why these bugs happen, but
part of it seems to be that depending on whether a pipelinespec is
embedded or not, sometimes the PipelineTask instances are v1beta1 and
sometimes v1alpha1. The logic to construct the graph is intentionally
using a "Task" interface instead of dealing with either of thse
directly, so it seems that in order to work properly, we need to have
the same logic for the Deps function for both. In the long run I hope we
can use exactly the same code instead of duplicating it, but in the
short term, now both functions are the same. To do this I had to make it
so that both v1alpha1 and v1beta1 could use the same Params and Results
types. I think this is a nice way to organize the code anyway, and
as a bonus, moving Params into a package outside of v1alpha1 and v1beta1
will make it so that we can more easily embed Conditions as part of #2079.

Fixes #2463

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [n/a] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

```
Fixes a bug which led to incorrect behavior when using results to order Tasks and embedding Pipeline specs.
```

### CODE REVIEW

File: pkg/apis/pipeline/v1beta1/pipeline_types.go [hunk 1]
```diff
@@ -152,6 +142,25 @@ func (pt PipelineTask) Deps() []string {
...
+	}
+	// Add any dependents from task results
+	for _, param := range pt.Params {
+		expressions, ok := results.GetVarSubstitutionExpressionsForParam(param)
+		if ok {
+			resultRefs := results.NewResultRefs(expressions)
+			for _, resultRef := range resultRefs {
+				deps = append(deps, resultRef.PipelineTask)
+			}
+		}
```
Comment (vdemeester) [inline pkg/apis/pipeline/v1beta1/pipeline_types.go:163]: This is the real fix, right ?

**Comment by bobcatfish** [comment]
Spoke with @vdemeester at length about this - gonna see if i can fix it without moving the params + results types into their own packages, but might still do that in a separate PR!

File: pkg/apis/pipeline/v1beta1/pipeline_types.go (hunk 1 above)
Comment (othomann) [reply pkg/apis/pipeline/v1beta1/pipeline_types.go:163]: This is the change to handle the dependencies.

File: pkg/apis/pipeline/v1alpha1/pipeline_conversion.go [hunk 2]
```diff
@@ -44,6 +44,7 @@ func (source *PipelineSpec) ConvertTo(ctx context.Context, sink *v1beta1.Pipelin
 	sink.Workspaces = source.Workspaces
 	sink.Description = source.Description
 	if len(source.Tasks) > 0 {
+		// todo: why is this failing?
```
Comment (bobcatfish) [inline pkg/apis/pipeline/v1alpha1/pipeline_conversion.go:47]: lol

**Pull request closed without merging**